### ISSUE TITLE
New feature: when a timer runs out, starts again where you left

### ISSUE THREAD

**Issue by bttfba** [Heckie75/kodi-addon-timers#8]
Suppose we are playing Media-1 (timed or not)
Media-2 has a timer set, for time xx:yy
When time xx:yy arrives, Media-1 stops and Media-2 starts
When Media-2 ends (or reach the time set for end), Media-1 should restart from the interrupted point.

I could also write the new feature in a different way, maybe including also what asked in a previous "new feature request":

When editing a timer for Media-A:
   - set the start time
   - set the end time
   - if timer interval (end - start) is equal or shorter than Media-A duration:
               - you are asked what to do at end of time interval:
                     - do nothing
                     - if Media-B is running when the timer starts, restart playing Media-B 
                     - exit, shutdown Kodi, ... whatelse is already there
    - if timer interval (end  -  start) is longer than Media-A duration:
                - you are asked what to do at end of Media-A:
                      - do nothing
                      - loop Media-A
                      - if Media-B is running when the timer starts, restart playing Media-B
                      - exit, shutdown Kodi, ... whatelse is already there

A bit more complex, but it would be possible to better manage the playout of several media.

As you can see, I'm a fan of TIMERS, trying to use it as a tool for managing a kind of private TV channel

**Comment by Heckie75** [comment]
Difficulty: medium
Estimation: 1 - 2 days (because of refactoring settings)

**Comment by bttfba** [comment]
Solution sounds good, thank you

**Comment by Heckie75** [comment]
Updated draft of new settings:
![grafik](https://user-images.githubusercontent.com/10388395/158334836-f9bd70f8-52d6-4566-85da-a1896a01b571.png)

**Comment by Heckie75** [comment]
@bttfba 

I have implemented the repeat / loop and resume functionality. Take a look at the new settings.

You can find it here: https://github.com/Heckie75/kodi-addon-timers/raw/master/script.timers.2.2.0-pre9.zip

You must probably restart kodi after updating the addon. 

Important: Maybe you want to backup your previous settings (see user data folders) since the settings are new and will be migrated. There is no backward compability in case that you reinstall an older version of the addon.

**Comment by Heckie75** [comment]
@bttfba 

Known issues so far:
* If length of scheduled item is shorter than timer interval then former item starts after scheduled item has fineshed but former item will be stopped after timer finished

**Comment by bttfba** [comment]
Ended first part of test: FLAG RESUME
Here the report:

**1. RESUME**
   1.1  End after TIME

      1.1.1  Length of scheduled video is longer than timed interval:
                   - scheduled video continues till end, then former video resumes from interrupted point
                   - test done with former video manually launched before the timer starts

      1.1.2  Length of scheduled video is longer than timed interval:
                   - scheduled video continues till end, then former video resumes from interrupted point
                   - test done with former video started with a timer earlier than the timer starts (first timer with flag 'Resume' either set or not set, no change)

      1.1.3  Length of scheduled video is shorter than timed interval:
                   - former video resumes at end of scheduled video from the interrupted point
                   - nothing happens when timed interval expires (just notification on the window)
                   - test done with former video manually launched before the timer starts

      1.1.4  Length of scheduled video is shorter than timed interval:
                   - former video resumes at end of scheduled video from the interrupted point
                   - nothing happens when timed interval expires (just notification on the window)
                   - test done with former video started with a timer earlier than the timer starts (first timer with flag 'Resume' either set or not set, no change)


   1.2  End after DURATION

       1.2.1  Length of scheduled video is longer than DURATION:
                   - scheduled video ends after 'DURATION' 
                   - formed video resumes from beginning (for a second) then jumps to interrupted point
                   - test done with former video manually launched before the timer starts

       1.2.2  Length of scheduled video is longer than DURATION:
                   - scheduled video ends after 'DURATION' 
                   - formed video resumes from beginning (for a second) then jumps to interrupted point
                   - test done with former video started with a timer earlier than the timer starts (first timer with flag 'Resume' either set or not set, no change)

       1.2.3  Length of scheduled video is shorter than DURATION:
                   - scheduled video ends (before expiration of the timer)
                   - former video resumes from interrupted point
                   - former video stops at expiration of the timer                            ( <<== known issue)
                   - test done with former video started manually launched before the timer starts

       1.2.4  Length of scheduled video is shorter than DURATION:
                   - scheduled video ends (before expiration of the timer)
                   - former video resumes from interrupted point
                   - former video stops at expiration of the timer                            ( <<== known issue)
                   - test done with former video started with a timer earlier than the timer starts (first timer with flag 'Resume' either set or not set, no change)

**Comment by bttfba** [comment]
Other tests done, with flag REPEAT
Here the report:

2. REPEAT
   2.1   No End (neither Time or Duration)

        2.1.1  Nothing playing
                   - set a timer on a video file with only start time set, no end, flag REPEAT set
                   - the video plays forever: when it reach the end, it starts again, and again ...    ( <<== issue, better to set a protection)

        2.1.2  Former video playing
                   - set a timer on a video file with only start time set, no end, flag REPEAT set
                   - the video plays forever: when it reachs the end, it starts again, and again ...    ( <<== issue, better to set a protection)

   2.2   End after TIME

        2.2.1  Length of scheduled video is longer than timed interval:
                   - scheduled video continues even after expiration of the timer (expiration notified)
                   - at end of the scheduled video, it starts again, and again, and again, apparently forever
                   - former video doesn't resume anymore (correct: flag RESUME not set)
                   - test done with former video manually launched before the timer starts

        2.2.2  Length of scheduled video is shorter than timed interval:
                   - scheduled video ends and restart always
                   - the timer expires, but nothing happens (expiration notified), scheduled video continues to play forever
                   - former video doesn't resume anymore  (correct: flag RESUME not set)
                   - test done with former video manually launched before the timer starts

        2.2.3  Length of scheduled video is longer than timed interval and RESUME is set:
                   - scheduled video continues even after expiration of the timer (expiration notified)
                   - at end of the scheduled video, it starts again, and again, and again, apparently forever
                   - former video doesn't resume anymore (even if flag RESUME set)
                   - test done with former video manually launched before the timer starts

        2.2.4  Length of scheduled video is shorter than timed interval and RESUME is set:
                   - scheduled video ends and restart always
                   - the timer expires, but nothing happens (expiration notified), scheduled video continues to play forever
                   - former video doesn't resume anymore  (even if flag RESUME set)
                   - test done with former video manually launched before the timer starts

   2.3  End after DURATION

        2.3.1  Length of scheduled video is longer than DURATION, RESUME not set and no former video playing:
                   - scheduled video ends after 'DURATION', i.e. before the end of the scheduled video
                   - no loop at all

        2.3.2  Length of scheduled video is longer than DURATION, RESUME not set and former video playing:
                   - scheduled video ends after 'DURATION', i.e. before the end of the scheduled video
                   - no loop at all
                   - former video not resumed (correct, flag RESUME not set)

        2.3.3  Length of scheduled video is longer than DURATION, RESUME set and former video playing:
                   - scheduled video ends after 'DURATION', i.e. before the end of the scheduled video
                   - no loop at all
                   - formed video resumes from beginning (for a second) then jumps to interrupted point (correct: flag RESUME set)

        2.3.4  Length of scheduled video is shorter than DURATION, RESUME not set and no former video playing:
                   - scheduled video ends, then start again
                   - when timer expires (after DURATION), the scheduled video ends
                   - no additional loops

        2.3.5  Length of scheduled video is shorter than DURATION, RESUME not set and former video playing:
                   - scheduled video ends, then start again
                   - when timer expires (after DURATION), the scheduled video ends
                   - no additional loops
                   - former video doesn't resume (correct: flag RESUME not set)

        2.3.6  Length of scheduled video is shorter than DURATION, RESUME set and former video playing:
                   - scheduled video ends, then start again
                   - when timer expires (after DURATION), the scheduled video ends
                   - no additional loops
                   - former video doesn't resume (wrong: flag RESUME is set)                  <<== issue (probably same as above)

**Comment by bttfba** [comment]
In summary:
- resume works well, but one issue when DURATION is longer that the scheduled video
- repeat should be more limited: 
        - if "end after" is set to nothing or to TIME, "repeat" should not be available
        - if "end after" is set to "duration", then:
                  - if "duration" is longer than the scheduled video, it should loop till the expiration of "duration"
                  - if "duration" is shorter than the scheduled video, no loop, the scheduled video should be stopped after "duration" and no loop then

Probably there are more cases to test, I'll do them when retesting for improvements, if any.

That's my proposal, hoping it is feasible without big effort
Thanks
Fabio

**Comment by Heckie75** [comment]
@bttfba 

Thanks for testing it so well. To be honest, I haven't expected the difficulty of this little project ;-) In the meantime I have spent almost all of my free time for testing and trying how solve findings with the resume functionality. It is even more difficult since the internal events that Kodi triggers, e.g. after the player has ended, seem to differ a little bit between media types such as PVR or mp3 playlist. That's why I have to implement a little state machine. 

I have checked in my latest version and did already some tests. The resume functionality looks quit good. Let's focus on this one first. So far I haven't tested the loop functionality. This would be step two. 

see https://github.com/Heckie75/kodi-addon-timers/blob/master/script.timers.2.2.0-pre10.zip

![grafik](https://user-images.githubusercontent.com/10388395/159109148-af375f47-0f52-473c-80cc-366b113c47d3.png)

My next step is to fix resuming playlist since at this moment after a timer ends the player only plays one single file but not the whole former playlist.

**Comment by bttfba** [comment]
@Heckie75 
You did a wonderful job and i have to thank you so much!
I agree: let's work on 'resume', i think it could be a nice feature to have this working appropriately and for now we can forget about 'repeat'. In addition, I discovered there is something like a 'loop' in the Player (slice menu, when you go with the mouse on the left) and it is working very badly (I also open a feature request in the main Kodi forum to ask for improvements). So probably we can survive with that and forget about looping inside Timers.
 
What is inside 'pre10' ? is there something new?

**Comment by Heckie75** [comment]
@bttfba 

I spent more time for the resume function esp. in combination with playlists so that the whole list continues and not only the former played single item. 

This is what's new in https://github.com/Heckie75/kodi-addon-timers/raw/master/script.timers.2.2.0-pre11.zip

Since I am few days on holiday I hadn't the time and also the equipment for testing it very well. Give it a try. If it's not working let me know.

BTW: I already know that resuming doesn't work with Zattoo PVR. But since this is very special I won't spent any effort on this. No idea if it works with TVHeadend since I was not able to test it. where I am.

**Comment by bttfba** [comment]
@Heckie75 
I'm taken too. Give me a couple of days to come back with test results. Anyway, don't waste your holidays for this! We can wait.
And also i agree on repeat: we can delay it after all! Playing Playlists and Folders + Resume is a great improvement!!
Thanks again
Fabio

**Comment by bttfba** [comment]
@Heckie75 
I retested here and there the new features and here are my results:
A. timed folders and playlists look fine, no issues
B. Resume with scheduled video shorter than DURATION: i started testing the known issue (called 1.2.3) and there is some problem still: basically at the end of the scheduled video, before expiration of timer, the scheduled video starts again, instead of the former video and it starts from the point where the former was interrupted. When the timer expires, finally, no actions, correctly. So, in summary, the right actions are made on the wrong video (the scheduled instead of the former).
Tested also on music playlists and the behavior looks the same
C. Resume when scheduled video is longer than DURATION: it was working before, now when scheduled video ends, the former video doesn't resume and no play anymore. Probably, also here the video to resume is not the correct one, so everything stops.

Take your time, don't rush on this - Thanks
Fabio

**Comment by bttfba** [comment]
@Heckie75
I did quite an extensive test and I didn't find any issue! For me it works just fine.
I started from the two issues 1.2.3 and 1.2.4 above and they are ok now. Then i did some tests with the flag "resume" on or off in several conditions (former / scheduled videos overlapping in many ways).
Finally i tried what will actually be my main expected working mode: folder's files playing (activated by a timer) and a movie starting, due to a timer too, coming down in streaming from an external (internet) source. It works perfectly!
Then I did the same with some music files playing and a movie started in the middle and this also worked fine. 
So, for what i saw, version "2.2.0-pre14" is approved!

Now, up to you to manage this, but I would remove the "repeat" flag and I would release the pre14 version.
About "repeat", I don't know, I think from my point of view, I'm happy with those changes and looping could be delayed or even avoided. Up to you, depending on your available time.

I really thank you for you availability and speed! It was really nice working with you
Ciao!!
Fabio

**Comment by Heckie75** [comment]
@bttfba 

That sounds good. My tests were also fine so far. I am going to take a look at the repeat functionality, but if it is to difficult I am going to skip the repeat functionallity for now. The pull request is already pretty gib and the guys from the Kodi team must be able to review the changes.

However, last question in terms of the resume functionality. I have implemented that resuming is the default like this:
* set quick timer via context menu: resume is activated (default media action is "start and stop")
* set timer via context menu: resume is activated (default media action is "start and stop")
* set sleep timer: resume is deactivated (default media action is "stop at end")
* set snooze timer: resume is deactivated (default media action is "start at end" or if it is set from epg "start at begin" )

Keep also in mind that resuming will (should - must be tested again) be deactivated if user explicitly presses stop.  

I guess that snooze and sleep are clear. Do you agree default setting of quick timer and timer or should it be deactivated by default.

**Comment by bttfba** [comment]
@Heckie75 I agree with the default setting, sounds good.
The only concern is about some interference among timers, if launched in some nested or overlapping mode.
Anyway, I did already some test on this, as said, and everything seems working well

**Comment by Heckie75** [comment]
> The only concern is about some interference among timers, if launched in some nested or overlapping mode.

At this moment I am very happy since timer and resuming work.

Some nested timers can optimized:

Szenario 1: T2 nested in T1
```

T2 nested in T1

M1        |------->
T1                 |-------------------|
T2                      |---------|
    
Acual:    |---M1---|-T1-|---T2----|-T1-|----nothing---->
Expected  |---M1---|-T1-|---T2----|-T1-|--------M1----->
```


Szenario 2: T1 overlaps T2:
```
Test in progresss
```

**Comment by bttfba** [comment]
@Heckie75 
Sorry for this long silence, i had some family issues
I see you worked more and produced pre16 and pre17. Moreover pre15 is not available anymore.
Let me know if there is something new to test, i'd be available now.

By the way, I agree with your decisions described here above (resume in expert view is fine, default 'no resume' is also correct, i think.
Let's see if there are additional things to test
Thanks
Fabio

**Comment by Heckie75** [comment]
@bttfba 

Yes, I have really worked hard and it resulted in heavy refactoring. I have just checked in pre18, see
https://github.com/Heckie75/kodi-addon-timers/raw/master/script.timers.2.2.0-pre18.zip

It is the fist release that worked after the refactoring. I haven't tested everything yet. It is very likely that there are here and there hard exceptions. I am going to test it this week.

**Comment by bttfba** [comment]
@Heckie75 
I see pre19 is no more there and pre20 is not yet available ... let me know when ready, i'll try to test with different sources
The tests you did are great! and most of them are successful, so great job!
Thanks!

**Comment by Heckie75** [comment]
@bttfba I am going to fix these issues that are left and re-test all TC 4.x

I'll let you know if there is a new version from which I think that all scenarios are alright. Since I haven't tested playlists it would be great if you can test some of the scenarios.

**Comment by bttfba** [comment]
@Heckie75 sure, I'll test them as soon as you deliver the new version

**Comment by Heckie75** [comment]
@bttfba 

I have fixed these two issues and tested changes successfully. There is a new version,  see https://github.com/Heckie75/kodi-addon-timers/raw/master/script.timers.2.2.0-pre21.zip

Would be nice if you can support testing. Let me know what you have tested and it would be great if you can documenate your tests in a simular way I did also in case that these were successful.

**Comment by bttfba** [comment]
@Heckie75 
I'm having an error when loading the new version - see attached logger
I also tried to uninstall completely TIMERS and then loading again from zip file  --> same error
[kodi - Copy.log](https://github.com/Heckie75/kodi-addon-timers/files/8447097/kodi.-.Copy.log)

**Comment by bttfba** [comment]
Cleaned everything, unistalled Kodi and personal areas and started again from scratch - same error
See the log attached
[kodi.log](https://github.com/Heckie75/kodi-addon-timers/files/8447187/kodi.log)

**Comment by Heckie75** [comment]
Strange, may be it is caused by a different python version. I have an idea.

Just for documentation. this is the change

```
def get_now() -> tuple[datetime.time, datetime.timedelta]:
```

```
def get_now() -> tuple['datetime.time', 'datetime.timedelta']:
```

May be you wait for a new version. Testing is very hard and after a fix on the one side there is a bug on the other. I think about introducing automated tests.

**Comment by bttfba** [comment]
@Heckie75 
it seems that   _**tuple()**_  is a data type definition of type '_**type**_', so probably it is a matter of different syntax to use
I'm not an expert of Python, so really I don't know what this is used for and how it should be written

**Comment by bttfba** [comment]
@Heckie75 
Your change (adding ' outside the parameters in the tuple definition) is not fixing the issue, I got same error. So, 
I tried several changes and finally I found a way that is giving me no errors:
changed all lines where 'tuple' or 'list' are present, going from    **tuple[ aaa , bbb ]**   to   **tuple ( [ aaa , bbb ] )**
and the same for all the 'list'  .  Basically, added a round bracket outside the squared bracket.
They were a lot, in several files: each time I fixed one, I got an error on the next used.

Doing this, I got no errors at all and the TIMERS work (just a quick test on running a timer on a folder, no other test)

Now up to you to explain the change, and of course to implement or not it. Let me know what you think of this
Ciao
Fabio

**Comment by Heckie75** [comment]
@bttfba 
I think that this tuple / list problem is related to Windows. However, I checked in a new Version pre21 that work on my my Windows setup.

You don't need to test anything at this moment. You can see the latest status in my testcases. I learnt that anytime I fix one issue another is raised I'm going to introduce unit tests. After I have written all these test cases, I will try to fix the open issues.

**Comment by Heckie75** [comment]
@bttfba 
I have covered all scenarios with automated unittest. This helped closing the last gaps.

I have only re-tested the testcases manually which have failed earlier. May be you would like to test the latest version, i.e. https://github.com/Heckie75/kodi-addon-timers/raw/master/script.timers.2.2.0-pre22.zip

You can see all scenarios and testcases here: https://github.com/Heckie75/kodi-addon-timers#overview-of-supported-timer-scenarios

**Comment by bttfba** [comment]
@Heckie75 
tested few TC (two are new), all with success
See attached my file with details. Only a couple of notes added
[test_220412_pre22.txt](https://github.com/Heckie75/kodi-addon-timers/files/8480869/test_220412_pre22.txt)

**Comment by Heckie75** [comment]
@bttfba
Many thanks for testing. According to resuming of an enclosing timers I think that it only restarts at correct position if the global setting "Seek to correct time if timer starts belatedly" is activated. This is because internally this scenario is handled identically. 

I have also tested the recent days on Linux, Windows and Android. I found another bug related to snooze context menu but now it works fine. 

Since the scheduler is more or less a complete rewrite I'll call this version 3.0.0. I am going to send the pull request right now.
